source/credentials_provider_imds.c

```diff
@@ -155,6 +155,18 @@ static struct imds_provider_user_data *s_imds_provider_user_data_new(
 static void s_on_get_credentials(const struct aws_credentials *credentials, int error_code, void *user_data) {
     (void)error_code;
     struct imds_provider_user_data *wrapped_user_data = user_data;
+    if (error_code == AWS_OP_SUCCESS) {
+        AWS_LOGF_INFO(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "id=%p: IMDS provider successfully retrieved credentials",
+            (void *)wrapped_user_data->imds_provider);
+    } else {
+        AWS_LOGF_ERROR(
```

Comment:
I'd prefer this to be info level.  The imds provider is added in non-ecs environments unless explicitly disabled, and making this error level means error-tagged log lines will show up any time an application is run outside of ec2.  Sometimes that's unavoidable, but in general we want to avoid flagging something at error level when it's an optional thing that will fail under ordinary circumstances.

Other than that, looks great.

Reply:
In practice, if we ever get to the point of running the IMDS provider in one of these environments, we're going to spew a bunch of error-level logs anyway. Actually, that's why I didn't notice this — we have the CRT's error-level logs turned off by default in Mountpoint clients because they are so spammy.

Here's what it looks like trying to hit IMDS on my laptop:
```
2024-03-22T20:17:11.926165Z ERROR awscrt::AuthCredentialsProvider: Failed to resolve role arn during sts web identity provider initialization.
2024-03-22T20:17:13.130559Z ERROR awscrt::AuthCredentialsProvider: (id=0x6000039dc1c0) Failed to source credentials from running process credentials provider with command: ada  credentials print --profile=dev 2>&1, err:The specified header was not found
2024-03-22T20:17:13.137216Z ERROR awscrt::channel-bootstrap: id=0x6000007d8230: Last attempt failed to create socket with error 1049
2024-03-22T20:17:13.137263Z ERROR awscrt::http-connection: static: Client connection failed with error 1049 (AWS_IO_SOCKET_NO_ROUTE_TO_HOST).
2024-03-22T20:17:13.137292Z  WARN awscrt::connection-manager: id=0x13b7063f0: Failed to obtain new connection from http layer, error 1049(socket connect failure, no route to host.)
2024-03-22T20:17:13.137314Z  WARN awscrt::connection-manager: id=0x13b7063f0: Failed to complete connection acquisition with error_code 1049(socket connect failure, no route to host.)
2024-03-22T20:17:13.137331Z  WARN awscrt::Unknown: id=0x13b7062d0: IMDS Client failed to acquire a connection, error code 1049(socket connect failure, no route to host.)
2024-03-22T20:17:13.137363Z ERROR awscrt::Unknown: (id=0x13b7062d0) IMDS client failed to update the token from IMDS.
2024-03-22T20:17:13.248366Z ERROR awscrt::channel-bootstrap: id=0x6000007d8230: Last attempt failed to create socket with error 1052
2024-03-22T20:17:13.248444Z ERROR awscrt::http-connection: static: Client connection failed with error 1052 (AWS_IO_SOCKET_NOT_CONNECTED).
2024-03-22T20:17:13.248465Z  WARN awscrt::connection-manager: id=0x13b7063f0: Failed to obtain new connection from http layer, error 1052(socket not connected.)
2024-03-22T20:17:13.248488Z  WARN awscrt::connection-manager: id=0x13b7063f0: Failed to complete connection acquisition with error_code 1052(socket not connected.)
2024-03-22T20:17:13.248506Z  WARN awscrt::Unknown: id=0x13b7062d0: IMDS Client failed to acquire a connection, error code 1052(socket not connected.)
2024-03-22T20:17:13.248524Z ERROR awscrt::AuthCredentialsProvider: id=0x6000039dc500: IMDS provider failed to retrieve role: socket not connected.
2024-03-22T20:17:13.250612Z ERROR awscrt::AuthCredentialsProvider: (id=0x6000039dc180) Default chain credentials provider failed to source credentials with error 6153(aws-c-auth: AWS_AUTH_CREDENTIALS_PROVIDER_IMDS_SOURCE_FAILURE, Valid credentials could not be sourced by the IMDS provider)
2024-03-22T20:17:13.250865Z ERROR awscrt::AuthSigning: (id=0x600000fd0320) Credentials Provider failed to source credentials with error 6153(aws-c-auth: AWS_AUTH_CREDENTIALS_PROVIDER_IMDS_SOURCE_FAILURE, Valid credentials could not be sourced by the IMDS provider)
2024-03-22T20:17:13.250891Z ERROR awscrt::S3MetaRequest: id=0x13b709080 Meta request could not sign HTTP request due to error code 6146 (Attempt to sign an http request without credentials)
2024-03-22T20:17:13.250910Z ERROR awscrt::S3MetaRequest: id=0x13b709080 Could not prepare request 0x15b605670 due to error 6146 (Attempt to sign an http request without credentials).
2024-03-22T20:17:13.253796Z  WARN list_objects{id=0 bucket="bornholt-test-bucket" continued=false delimiter="" max_keys="500" prefix=""}: mountpoint_s3_client::s3_crt_client: duration=1.32434725s request_id=<unknown> error=ClientError(NoSigningCredentials) meta request failed
```

That said, I'll change it in the interest of at least not making the spam problem worse :-)